bot/exts/filtering/_settings.py

```diff
@@ -0,0 +1,232 @@
+            validation_data[entry_name] = entry_data
+        elif entry_name not in _already_warned:
+            log.warning(
+                f"A setting named {entry_name} was loaded from the database, but no matching class."
+            )
+            _already_warned.add(entry_name)
+    if defaults is None:
+        default_actions = None
+        default_validations = None
+    else:
+        default_actions, default_validations = defaults
+    return (
+        ActionSettings.create(action_data, defaults=default_actions, keep_empty=keep_empty),
+        ValidationSettings.create(validation_data, defaults=default_validations, keep_empty=keep_empty)
+    )
+
+
+class Settings(FieldRequiring, dict[str, T]):
+    """
+    A collection of settings.
+
+    For processing the settings parts in the database and evaluating them on given contexts.
+
+    Each filter list and filter has its own settings.
+
+    A filter doesn't have to have its own settings. For every undefined setting, it falls back to the value defined in
+    the filter list which contains the filter.
+    """
+
+    entry_type = T
+
+    _already_warned: set[str] = set()
+
+    @abstractmethod
+    def __init__(self, settings_data: dict, *, defaults: Settings | None = None, keep_empty: bool = False):
+        super().__init__()
+
+        entry_classes = settings_types.get(self.entry_type.__name__)
+        for entry_name, entry_data in settings_data.items():
+            try:
+                entry_cls = entry_classes[entry_name]
+            except KeyError:
+                if entry_name not in self._already_warned:
+                    log.warning(
+                        f"A setting named {entry_name} was loaded from the database, "
+                        f"but no matching {self.entry_type.__name__} class."
+                    )
+                    self._already_warned.add(entry_name)
+            else:
+                try:
+                    entry_defaults = None if defaults is None else defaults[entry_name]
+                    new_entry = entry_cls.create(
+                        entry_data, defaults=entry_defaults, keep_empty=keep_empty
+                    )
+                    if new_entry:
+                        self[entry_name] = new_entry
+                except TypeError as e:
+                    raise TypeError(
+                        f"Attempted to load a {entry_name} setting, but the response is malformed: {entry_data}"
+                    ) from e
+
+    @property
+    def overrides(self) -> dict[str, Any]:
+        """Return a dictionary of overrides across all entries."""
+        return reduce(operator.or_, (entry.overrides for entry in self.values() if entry), {})
+
+    def copy(self: TSettings) -> TSettings:
+        """Create a shallow copy of the object."""
+        copy = self.__class__({})
+        copy.update(super().copy())  # Copy the internal dict.
+        return copy
```

Comment:
Is this different from `copy.copy(self)`?

Reply:
No, it is not 👍

Reply:
12e1e37